PYI-698: Accept JWT client authentication

Wynndow opened the pull request:
**Config PR to support these changes: https://github.com/alphagov/di-ipv-config/pull/163**
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Adds a new service to validate client authentication JWTs in the token
endpoint.

Also refactors some services and injects them to make things easier to
test.

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYI-698](https://govukverify.atlassian.net/browse/PYI-698)

Wynndow committed:
PYI-698: Accept JWT client authentication

Adds a new service to validate client authentication JWTs in the token
endpoint.

Also refactors some services and injects them to make things easier to
test.

Wynndow committed:
PYI-698: Use ClientAuthenticationVerifier for auth JWT verification

Use the client authentication verifier supplied in the Nimbus library.

This will save having to use a bunch of custom validation code, and
means it will be easier to have a standardised approach across system
components.

We need to implement the `ClientCredentialsSelector` class, so that we
can tell the verifier how to fetch the public key for each client. The
implementation added here just pulls the certificate from the config
class we're using. Other implementations, such as in core-back, could
pull them directly from the param store.

dannyhayes-gds commented on di-ipv-credential-issuer-stub/src/main/java/uk/gov/di/ipv/stub/cred/auth/ClientJwtVerifier.java:
```diff
@@ -0,0 +1,71 @@
+        }
+    }
+
+    private Map<String, List<String>> listifyParamValues(QueryParamsMap requestParams) {
```
We might not need this method. The PrivateKeyJWT.parse(queryParams) has a few different overloaded methods. One of them accepts just a single string value which are the query params i.e "var1=hello&var2=bye&var=hello" etc...

So maybe instead of passing in the QueryMap object into the authenticateClient method, we can instead pass in the original query params as a string, which I think we can get from something like
request.queryString() 

Just a suggestion though, not too fussed if we want to keep it as is and build the map into the required Map<String, List<String>> format

Wynndow replied:
Nice. Have pushed a commit to use the query string version instead.

dannyhayes-gds commented on di-ipv-credential-issuer-stub/src/main/java/uk/gov/di/ipv/stub/cred/auth/ClientJwtVerifier.java:
```diff
@@ -0,0 +1,71 @@
+        }
+
+        try {
+            verifier.verify(authenticationJwt, null, null);
```
Do you know if this will test against the expiry ttl time being too far in the future? Or do we need to do that separately?

Wynndow replied:
I don't think it does. I think we agreed that for the stub we're not too concerned about this. I've spoken to @SunilDGDS about having to do it as an extra check in his core-back verifying PR though.

dannyhayes-gds replied:
Ah okay thats sounds good 👍

Wynndow committed:
PYI-698: Use query string with JWT verifier

The parse method for PrivateKeyJWTs is overloaded to allow a query
string, which we can just get from the request. Saves silly data
munging methods.

dannyhayes-gds reviewed (approved):
Looks good 👍
